MetaMask/metamask-extension
# Fix conversion rate issues

Fixes: #9492

Explanation:  In cases where a user has not entered a symbol ticker for a custom RPC network (or our conversion rate API doesn't recognize the asset), don't currently update our conversionRate in state and use the conversion rate from the previously used network with a valild conversion rate. Moreover, in useCurrencyDisplay (where we apply the conversion for most UI elements where the value of a secondary currency is required) we haven't had a clear conditional separation of values that require conversion (Secondary currencies / Fiat) and those that do not (Primary Currency / native assets).

 This PR, in tandem with this [one](https://github.com/MetaMask/controllers/pull/465) allows conversionRates to be null when we cannot calculate them. In such cases we do not show a converted secondary currency (even if the user has selected to show conversions in testnets). This prevents us from showing inaccurate conversions. It also adds logic to avoid applying conversion rates to primary currencies which results in some of the more alarming bugs (show in video in linked PR above).

tldr: when reviewing start at useCurrencyDisplay.js
 
 Additional work here:
- Removes user-preferenced-token-input & user-preferenced-currency-input components because they are just wrapping token-input and currency-input and passing a single additional prop (useNativeCurrencyAsPrimaryCurrency) that can be added by the component's containers. IMO this pattern makes it harder to tell where props are coming from and how to reason about what the components are trying to achieves. Furthermore there is currently no where that the token-input or currency-input is used where it wasn't wrapped by user-preference-token-input /user-preference-currency-input so the pattern isn't creating any meaningful distinction between the wrapped and unwrapped versions.
- add functionality such that when a user has selected fiat as their primary currency token asset list items also show fiat as primary.


Work to follow:
- Proposal to replace useCurrencyDisplay with a hook that is more agnostic and versatile to consolidate the many patterns we have for calculating and displaying `currentCurrency` and `nativeCurrency` together and clean up the impossible to reason about logic that lives in components like currency-input.component and token-input.component.
- A PR to change the names of `currentCurrency` and `nativeCurrency` to `preferredCurrency` and `networkCurrency` respectively per this [thread](https://consensys.slack.com/archives/G1L7H42BT/p1620939857402600). If you have thoughts on this please include comments here?
 
Manual testing steps:  
  - In advanced settings, turn `Show Conversion on Testnets` on.
  - Go between mainnet, a custom rpc network where you have added a ticker symbol and a custom rpc network where you haven't added a ticker symbol.
  - Check that conversions are correct for all assets across these different networks (for the network with no ticker symbol there should be no conversions).
